mlrun/datastore/azure_blob.py

```diff
@@ -53,36 +56,65 @@ def get_storage_options(self):
             connection_string=self._get_secret_or_env(
                 "AZURE_STORAGE_CONNECTION_STRING"
             ),
+            tenant_id=self._get_secret_or_env("AZURE_STORAGE_TENANT_ID"),
+            client_id=self._get_secret_or_env("AZURE_STORAGE_CLIENT_ID"),
+            client_secret=self._get_secret_or_env("AZURE_STORAGE_CLIENT_SECRET"),
+            sas_token=self._get_secret_or_env("AZURE_STORAGE_SAS_TOKEN"),
         )
 
     def upload(self, key, src_path):
-        # Need to strip leading / from key
-        blob_client = self.bsc.get_blob_client(container=self.endpoint, blob=key[1:])
-        with open(src_path, "rb") as data:
-            blob_client.upload_blob(data, overwrite=True)
+        if self.bsc:
+            # Need to strip leading / from key
+            blob_client = self.bsc.get_blob_client(container=self.endpoint, blob=key[1:])
+            with open(src_path, "rb") as data:
+                blob_client.upload_blob(data, overwrite=True)
+        else:
+            remote_path = f"{self.endpoint}{key}"
+            self._filesystem.put_file(src_path, remote_path)
 
     def get(self, key, size=None, offset=0):
-        blob_client = self.bsc.get_blob_client(container=self.endpoint, blob=key[1:])
-        size = size if size else None
-        return blob_client.download_blob(offset, size).readall()
+        if self.bsc:
+            blob_client = self.bsc.get_blob_client(container=self.endpoint, blob=key[1:])
+            size = size if size else None
+            return blob_client.download_blob(offset, size).readall()
+        else:
+            blob = self._filesystem.cat_file(key, start=0)
+            return blob
 
     def put(self, key, data, append=False):
-        blob_client = self.bsc.get_blob_client(container=self.endpoint, blob=key[1:])
-        # Note that append=True is not supported. If the blob already exists, this call will fail
-        blob_client.upload_blob(data, overwrite=True)
+        if self.bsc:
+            blob_client = self.bsc.get_blob_client(container=self.endpoint, blob=key[1:])
+            # Note that append=True is not supported. If the blob already exists, this call will fail
```

Comment:
I just noticed that this comment is not correct, as we're using `overwrite=True`, so it will not fail but rather overwrite. This is of course not your fault 😃. 
Still, I'd prefer in general for this method to blow up if `append=True` is provided since it may cause data loss to the user when expecting an append operation and getting an overwrite instead. 
If you could add this test, that's great. Otherwise I'll do that on a different PR.

Reply:
I just really thought about this operation.  It's worth noting that a BlockBlob (the default in Azure) does not support append operations at all.

Reply:
Yes, that's why it was omitted in the basic implementation - you need to use AppendBlob type (which is not always supported) to work with append operations. Of course, I'm less concerned about not supporting append - I'm concerned about not notifying the user that append in implicitly replaced with overwrite. That's why failing such a call seems appropriate.